### ISSUE TITLE
`pycbc_optimal_snr` produces null optimal SNRs when reading HDF injections and using more than one core

### ISSUE THREAD

**Issue by titodalcanton** [gwastro/pycbc#3693]

**Comment by titodalcanton** [comment]
The issue appears to arise in the final collection of the results from the workers, rather than in the actual calculation of the SNR itself.

**Comment by titodalcanton** [comment]
I am not sure why, but this assignment has no effect:
```python
setattr(inj, column, sval)
```
which leads to the optimal SNRs being left at their initialization value (zero).

**Comment by spxiwh** [comment]
It might be copying the `inj` object when splitting to multiple cores. We might need to return the SNRs as an array (or something) and then set values outside of the multiprocessing bit.

**Comment by titodalcanton** [comment]
Well, I am finding that
```python
setattr(inj, column, sval)
assert getattr(inj, column) == sval
```
fails, i.e. somehow the `inj` object is (silently) immutable *within the same thread*. But yes, a complete refactor could be one way around this issue.

**Comment by titodalcanton** [comment]
Fixed by #3724.
